pp/series_data/querier/instant_querier.h

```diff
@@ -13,20 +17,54 @@ class InstantQuerier {
-    if (storage.open_chunks.size() <= ls_id || storage.open_chunks[ls_id].is_empty()) [[unlikely]] {
-      return;
+  explicit InstantQuerier(DataStorage& storage) : storage_(storage) {}
+
+  template <typename LsIDStorage, typename SampleStorage>
+  void query(SampleStorage& samples, const LsIDStorage& label_set_ids, const Timestamp& timestamp) noexcept {
+    assert(std::size(samples) == std::size(label_set_ids));
+
+    for (auto&& [sample, ls_id] : std::ranges::views::zip(samples, label_set_ids)) {
+      query_sample(sample, ls_id, timestamp);
     }
-    if (const auto series_last_ts = Decoder::get_series_max_timestamp(storage, ls_id); timestamp >= series_last_ts) {
-      sample = {.timestamp = series_last_ts, .value = Decoder::get_open_chunk_last_value(storage, storage.open_chunks[ls_id])};
-    } else if (const auto series_first_ts = Decoder::get_series_min_timestamp(storage, ls_id); timestamp >= series_first_ts) {
-      check_inside_series(sample, storage, ls_id, timestamp);
+  }
+
+  template <typename LsIDStorage, typename SampleStorage>
+  void query_reload(SampleStorage& samples, const LsIDStorage& label_set_ids, const Timestamp& timestamp) noexcept {
+    assert(std::size(samples) == std::size(label_set_ids));
+
+    for (auto&& [sample, ls_id] : std::ranges::views::zip(samples, label_set_ids)) {
+      if (series_to_load_.is_set(ls_id)) {
+        query_sample(sample, ls_id, timestamp);
+        storage_.queried_series_bitmap.set_atomic(ls_id);
+      }
     }
   }
 
+  [[nodiscard]] bool need_loading() const noexcept { return series_to_load_.empty() == false; }
+  [[nodiscard]] const BareBones::Bitset& get_series_to_load() const noexcept { return series_to_load_; }
+
  private:
-  static void check_inside_series(Sample& sample, const DataStorage& storage, LabelSetID ls_id, const Timestamp& timestamp) noexcept {
-    for (const auto& chunk_data : DataStorage::SeriesChunks(&storage, ls_id)) {
+  DataStorage& storage_;
+  BareBones::Bitset series_to_load_;
+
+  PROMPP_ALWAYS_INLINE void query_sample(Sample& sample, LabelSetID ls_id, const Timestamp& timestamp) noexcept {
+    if (storage_.open_chunks.size() <= ls_id || storage_.open_chunks[ls_id].is_empty()) [[unlikely]] {
+      return;
+    }
+    if (const auto series_last_ts = Decoder::get_series_max_timestamp(storage_, ls_id); timestamp >= series_last_ts) {
+      sample = {.timestamp = series_last_ts, .value = Decoder::get_open_chunk_last_value(storage_, storage_.open_chunks[ls_id])};
+    } else if (const auto series_first_ts = Decoder::get_series_min_timestamp(storage_, ls_id); timestamp >= series_first_ts) {
+      if (storage_.unloaded_series_bitmap.is_set(ls_id)) {
+        series_to_load_.set(ls_id);
+        return;
+      }
+      check_inside_series(sample, ls_id, timestamp);
+      storage_.queried_series_bitmap.set_atomic(ls_id);
```

Comment:
```suggestion
      storage_.queried_series_bitmap.set_atomic(ls_id);
      if (storage_.unloaded_series_bitmap.is_set(ls_id)) {
        series_to_load_.set(ls_id);
        return;
      }
      check_inside_series(sample, ls_id, timestamp);
```
This allow remove `storage_.queried_series_bitmap.set_atomic(ls_id);` from `query_reload`